Update bcm2711_bootloader_config.md

timg236 opened the pull request:
* Use backticks to format values which are likely to be copied and pasted to avoid people copying the double quotes as well
* Simplify the editing section at the start to remove information duplicated on the booteeprom page. Just present the simple '--edit' option
* Tablify some enums and link to usbboot repo for CM4

lurch commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 
 * Boot mode `0x0` will retry the SD boot if the SD card detect pin indicates that the card has been inserted or replaced.
 * The default boot order is `0xf41` which means continuously try SD then USB mass storage.
+* USB device boot is intended for use with Compute Module 4 to load a debug image e.g. a Linux RAM disk. This should be the last boot option because it does not currently support timeouts or retries.
```
Is "debug image" the correct phrasing here? Isn't USB device boot also used to flash the EMMC? (which is something that most users wouldn't class as "debug")

timg236 replied:
specified that this a custom image i.e. something other than the usbboot MSD device mode feature

lurch commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
+|  0x0  |  NONE     |  Stop and display error pattern                                          |
+|  0x1  |  SD CARD  |  SD card (or EMMC on Compute Module 4)                                   |
+|  0x2  |  NETWORK  |  Network boot                                                            |
+|  0x3  |  USB DEV  |  USB device boot - See [usbboot](https://github.com/raspberrypi/usbboot) |
```
Is it worth adding the version-date in which USB DEV was added? :man_shrugging:

timg236 replied:
Picked 2020-09-03 ... close enough!

lurch commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 The BOOT_ORDER setting allows flexible configuration for the priority of different bootmodes. It is represented as 32bit unsigned integer where each nibble represents a bootmode. The bootmodes are attempted in lowest significant nibble to highest significant nibble order.
 
-E.g. 0x21 means try SD first followed by network boot then stop. Whereas 0x2 would mean try network boot and then stop without trying to boot from the SD card.
+E.g. `0x21` means try SD first followed by network boot then stop. Whereas 0x2 would mean try network boot and then stop without trying to boot from the SD card.
```
Backticks around 0x2 ?

timg236 replied:
done

lurch commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
-Minimum: 5000  
-Default: 45000  
+Minimum: `5000`  
+Default: `45000` 
```
You've accidentally changed the two-spaces at the end of this line to a single-space, which means the forced line-break goes wrong.

lurch commented:
Minor comments, but otherwise this LGTM :+1:

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -1,40 +1,15 @@
-# Pi 4 Bootloader Configuration
+# Raspberry Pi 4 Bootloader Configuration
```
Style guide (https://github.com/raspberrypilearning/style-guide/blob/master/style-guide.md) says only first word (and product names) should be capitalised, so this becomes:

`Raspberry Pi 4 bootloader configuration`

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 
 If 1 then enable UART debug output on GPIO 14 and 15. Configure the receiving debug terminal at 115200bps, 8 bits, no parity bits, 1 stop bit.
```
`1` in backticks.

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 
 ### WAKE_ON_GPIO
 
 If 1 then 'sudo halt' will run in a lower power mode until either GPIO3 or GLOBAL_EN are shorted to ground.
```
`1` in backticks.

andrum99 replied:
Also, the meaning could do with some further explanation, e.g.:

`If 1 then once the operating system shuts down (e.g. from sudo halt), the Raspberry Pi will run in a low-power mode until either GPIO3 or GLOBAL_EN are shorted to ground.`

timg236 replied:
I think it's self-explanatory

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 
 ### POWER_OFF_ON_HALT
 
 If 1 and WAKE_ON_GPIO=0 then switch off all PMIC outputs in halt. This is lowest possible power state for halt but may cause problems with some HATs because 5V will still be on. GLOBAL_EN must be shorted to ground to boot.
```
`1` in backticks.

andrum99 replied:
Also might want some further explanation of "halt", e.g.:

`If 1 and WAKE_ON_GPIO=0 then switch off all PMIC outputs when the Pi is shut down.`

And `WAKE_ON_GPIO=0` in backticks.

timg236 replied:
reworded

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 The BOOT_ORDER setting allows flexible configuration for the priority of different bootmodes. It is represented as 32bit unsigned integer where each nibble represents a bootmode. The bootmodes are attempted in lowest significant nibble to highest significant nibble order.
 
-E.g. 0x21 means try SD first followed by network boot then stop. Whereas 0x2 would mean try network boot and then stop without trying to boot from the SD card.
+E.g. `0x21` means try SD first followed by network boot then stop. Whereas `0x2` would mean try network boot and then stop without trying to boot from the SD card.
```
I think this would be easier to understand if the second sentence read as follows:

`Whereas 0x2 would mean try network boot only.` (with backticks round 0x2).

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 
 ### SD_BOOT_MAX_RETRIES
 The number of times that SD boot will be retried after failure before moving to the next boot mode defined by `BOOT_ORDER`.  
 -1 means infinite retries.
```
`-1` in backticks.

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -43,89 +18,92 @@ This section describes all the configuration items available in the bootloader.
 
 ### NET_BOOT_MAX_RETRIES
 The number of times that network boot will be retried after failure before moving to the next boot mode defined by `BOOT_ORDER`.  
 -1 means infinite retries.
```
`-1` in backticks.

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -260,31 +240,31 @@ This is intended to allow problematic (e.g. very slow to enumerate) devices to b
 
 ### XHCI_DEBUG
-This property controls the verbosity of USB debug messages for mass storage boot mode. Enabling all of these messages generates a huge amount of log data which will slow down booting and may even cause boot to fail. For verbose logs it's best to use `NETCONSOLE`.
+This property is a bit field controls the verbosity of USB debug messages for mass storage boot mode. Enabling all of these messages generates a huge amount of log data which will slow down booting and may even cause boot to fail. For verbose logs it's best to use `NETCONSOLE`.
```
Need e.g. `which` after `is a bit field`.

timg236 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -1,40 +1,15 @@
-# Example change. If you have a UART cable then setting BOOT_UART=1 will help debug boot issues
-BOOT_UART=1
+To view the current EEPROM configuration:
+`rpi-eeprom-config`
```
The double space after the configuration: has the desired effect of creating a new line

timg236 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -1,40 +1,15 @@
-
-If you update your bootloader via apt, then any configuration changes made using the process described here will be migrated to the updated bootloader.
+To edit it and apply the updates to latest EEPROM release:
+`sudo -E rpi-eeprom-config --edit`
```
done

timg236 commented:
Addressed all the minor comments and updated BOOT_ORDER to cover the more esoteric 0xe STOP and SD card polling modes. These were added before 0xf and are probably unused except for tests

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -1,131 +1,110 @@
-* 0xf - RESTART (loop) - start again with the first boot order field. (since 2020-09-03)
+| Value | Mode      | Description                                                                                       |
+|-------|-----------|---------------------------------------------------------------------------------------------------|
+|  0x1  |  SD CARD  |  SD card (or EMMC on Compute Module 4)                                                            |
```
I'm wondering if `or` should be removed, since CM4 only has eMMC. It is also possible to use SD card via GPIO on the CM4, but I'm not sure if you can boot from it.

andrum99 replied:
Also `EMMC` -> `eMMC`

timg236 replied:
CM4 has either eMMC or SD card on LITE. It's not currently possible to boot from SDHOST controller

lurch replied:
The CM4 Lite doesn't have eMMC, and requires the use of an external SD card.

andrum99 replied:
> The CM4 Lite doesn't have eMMC, and requires the use of an external SD card.

That's what I thought. So without working SD card boot on CM4, the word `or` should be removed.

andrum99 commented on hardware/raspberrypi/bcm2711_bootloader_config.md:
```diff
@@ -1,131 +1,110 @@
+|  0xe  |  STOP     |  Stop and display error pattern (since 2020-09-03). A power cycle is required to exit this state. |
+|  0xf  |  RESTART  |  Start again with the first boot order field. (since 2020-09-03)                                  |
+
+After trying each non-zero boot-mode the bootloader stops. However, from 2020-09-03 the bootloader will monitor the SD card-detect pin and try SD boot if a new SD card is inserted.
```
`boot-mode` -> `boot mode`

andrum99 replied:
`card-detect` -> `card detect`

lurch replied:
card-detect is a specific pin on the SD slot, so should possibly remain hyphenated.
